README.md
# Python VOHLC

Version 0.0.1

[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

Creates Volume-based candlesticks from time-based ohlcv data retrieved from cryptocurrency exchanges.

- Output can be either sqlite or csv (for version 0.0.1)
- Primarily just a utility for something that I'm working on
- Should not be very complex. Keep it simple, just do what it needs to do
- Does not keep the timestamps
- Potentially give the option to not write the output, just keep in memory

## Installation

You may install from source, or install from pypi:

```shell
pip install vohlc
```

[View on Pypi](https://pypi.org/project/vohlc/)

CHANGELOG.txt
Change Log
==========

0.0.1 (2021//)
------------------
- First Release

tests/test1.py
"""
1. Change the name from Configuration. Since that's the only import, make it meaningful. It's in
a cile called client, so something like that.
2. Change the persistence configuration to be more intuitive. I don't like having to pass everything
in this way, especially with the chaining, that just gets too big.
3. Test the persistence functionality, other than that it works great.
Definitely a program that will take some time to run depending on timeframe and start date.
"""
import os
from vohlc import RunBasic, RunCsv, RunSqlite


df = RunBasic('coinbasepro', 1000).collect('BTC/USD', '1d', '2021-01-01 00:00:00') \
    .collect_data().transform_data()

db_path = os.path.join(os.getcwd(), 'data', 'btcusd.db')
csv_path = os.path.join(os.getcwd(), 'data', 'btcusd.csv')

print(db_path)
print(csv_path)

print(df)
sym = 'BTC/USD'
tf = '1d'
iso = '2021-01-01 00:00:00'

csv_df = RunCsv('coinbasepro', 1000).collect(sym, tf, iso, csv_path)
sqlite_df = RunSqlite('coinbasepro', 1000).collect(sym, tf, iso, db_path, 'btc_usd_data')


vohlc/client.py
import sqlite3 as sql
import pandas as pd
from .base.collect_data import create_exchange, run_ohlcvs_collection
from .base.transform_ohlcv import batch_df
from typing import List, Dict
from enum import Enum


class Persistence(Enum):
    NONE = 0
    CSV = 1
    SQLITE = 2


class Transform:
    def __init__(self, dataframe: pd.DataFrame, batch_size: int):
        self.df = dataframe
        self.batch = batch_size

    def transform_data(self) -> pd.DataFrame:
        return batch_df(self.df, self.batch)


class WriteCsv:
    def __init__(self, filepath: str, dataframe: pd.DataFrame):
        self.filepath = filepath
        self.df = dataframe

    def write(self):
        self.df.to_csv(self.filepath, index=False)


class WriteSql:
    def __init__(self, db_filepath, tablename, data: pd.DataFrame):
        self.db_file = db_filepath
        self.table = tablename
        self.df = data

    def write(self):
        con = sql.connect(self.db_file)
        self.df.to_sql(name=self.table, con=con, index=False)
        con.close()


class Collect:
    def __init__(self, exchange, symbol, timeframe, from_iso, batch_size):
        self.exchange = exchange
        self.symbol = symbol
        self.timeframe = timeframe
        self.from_iso = from_iso
        self.batch_size = batch_size

    def collect_data(self) -> Transform:
        df = run_ohlcvs_collection(self.exchange, self.symbol, self.timeframe, self.from_iso)
        return Transform(df, self.batch_size)


class RunBasic:
    def __init__(self, exchangeid: str, batch_size: int, rate_limit=600):
        self.exchange = create_exchange(exchangeid, rate_limit)
        self.batch_size = batch_size

    def collect(self, symbol, timeframe, from_iso) -> Collect:
        return Collect(self.exchange, symbol, timeframe, from_iso, self.batch_size)


class RunSqlite:
    def __init__(self, exchangeid: str, batch_size: int, rate_limit=600):
        self.exchange = create_exchange(exchangeid, rate_limit)
        self.batch_size = batch_size

    def collect(self, symbol, timeframe, from_iso, db_filepath, tablename):
        _collect = Collect(self.exchange, symbol, timeframe, from_iso, self.batch_size)
        data = _collect.collect_data().transform_data()
        WriteSql(db_filepath, tablename, data).write()
        return data


class RunCsv:
    def __init__(self, exchangeid: str, batch_size: int, rate_limit=600):
        self.exchange = create_exchange(exchangeid, rate_limit)
        self.batch_size = batch_size

    def collect(self, symbol, timeframe, from_iso, csv_filepath):
        _collect = Collect(self.exchange, symbol, timeframe, from_iso, self.batch_size)
        data = _collect.collect_data().transform_data()
        WriteCsv(csv_filepath, data).write()
        return data


vohlc/base/transform_ohlcv.py
import pandas as pd
from typing import List


def vohlc_cols() -> List[str]:
    return ['open', 'high', 'low', 'close']


def parse_ohlc(df: pd.DataFrame) -> pd.DataFrame:
    op = float(df.loc[0, 'open'])
    last_idx = df.index[-1]
    cl = float(df.loc[last_idx, 'close'])
    max_ = df.high.max()
    min_ = df.low.min()
    return pd.DataFrame(
        columns=vohlc_cols(),
        data=[[op, max_, min_, cl]]
    )


def batch_df(df: pd.DataFrame, batch_size: int):
    df['volume'] = df['volume'].astype(float)
    df['total'] = df['volume'].cumsum()
    df['batch'] = (df['total'] / batch_size).astype(int)
    grouped_df = df.groupby('batch')
    ohlc_dfs = []
    for d in grouped_df.groups.keys():
        _df = grouped_df.get_group(d)
        _df.reset_index(inplace=True, drop=True)
        ohlc_df = parse_ohlc(_df)
        ohlc_dfs.append(ohlc_df)
    master_df = pd.concat(ohlc_dfs)
    master_df.reset_index(inplace=True, drop=True)
    return master_df


vohlc/base/collect_data.py
import ccxt
import pandas as pd
from datetime import datetime, timezone
from typing import List


def create_exchange(exchangeid, rate_limit=600):
    return getattr(ccxt, exchangeid)({
        'enableRateLimit': True,
        'rateLimit': rate_limit
    })


def start_time() -> int:
    today = datetime.utcnow().date()
    start_dt = datetime(today.year, today.month, today.day, tzinfo=timezone.utc)
    return int(start_dt.timestamp() * 1000)


def ohlcv_col_names() -> List[str]:
    return ['ts', 'open', 'high', 'low', 'close', 'volume']


def create_df(ohlcvs) -> pd.DataFrame:
    return pd.DataFrame(
        columns=ohlcv_col_names(),
        data=ohlcvs
    )


def update_df(dataframe: pd.DataFrame, ohlcvs) -> pd.DataFrame:
    df_2 = create_df(ohlcvs)
    df_ = pd.concat([dataframe, df_2])
    return df_.copy()


def ohlcvs_recursive(exchange, symbol, timeframe, end_timestamp, from_timestamp, dataframe=None) -> pd.DataFrame:
    if from_timestamp < end_timestamp:
        if dataframe is None:
            ohlcvs = exchange.fetch_ohlcv(symbol, timeframe, from_timestamp)
            df = create_df(ohlcvs)
            try:
                from_timestamp_ = ohlcvs[-1][0]
            except IndexError:
                return df
            return ohlcvs_recursive(exchange, symbol, timeframe, end_timestamp, from_timestamp_, df)
        else:
            ohlcvs = exchange.fetch_ohlcv(symbol, timeframe, from_timestamp)
            df = update_df(dataframe, ohlcvs)
            try:
                from_timestamp_ = ohlcvs[-1][0]
            except IndexError:
                return df
            return ohlcvs_recursive(exchange, symbol, timeframe, end_timestamp, from_timestamp_, df)
    else:
        return dataframe


def run_ohlcvs_collection(exchange, symbol, timeframe, from_iso) -> pd.DataFrame:
    from_timestamp = exchange.parse8601(from_iso)
    end_timestamp = start_time()
    return ohlcvs_recursive(exchange=exchange, symbol=symbol, timeframe=timeframe, end_timestamp=end_timestamp,
                            from_timestamp=from_timestamp)
